build/templates/_grpc_stub_interpreter.py/initialization_method.py.mako

```diff
@@ -15,9 +15,13 @@
 %>\
 
     def ${full_func_name}(${method_decl_params}):  # noqa: N802
+        metadata = (
+            ('x-api-key', self._grpc_options.api_key),
+        )
         ${capture_response}self._invoke(
```

Comment:
It looks like the API key is only used to initialize a session. After the session is initialized, another client can access the session without the API key. Is that the intent?

This might be acceptable for licensing but it would not be acceptable for authentication.

Reply:
You should strongly consider using an interceptor to specify metadata.

Reasons:
- An interceptor would automatically add metadata for all calls.
- You wouldn't have to modify any call sites, which would make this a lower-touch change. Instead, you would instantiate an interceptor object, call `grpc.intercept_channel(channel, interceptor)`, and use the channel that it returns.

We have been using interceptors to add metadata in C#. It's really easy with that API because it has a metadata interceptor built in and it provides an overload of `ChannelBase.Intercept()` that takes a `Func<Metadata, Metadata>`. I don't know if the gRPC Python API provides a built-in metadata interceptor, but writing a metadata interceptor for unary-unary calls should not be complicated.

Reply:
Here is an example of a "header adder interceptor".
https://github.com/grpc/grpc/blob/2d4f3c56001cd1e1f85734b2f7c5ce5f2797c38a/examples/python/interceptors/headers/header_manipulator_client_interceptor.py

Reply:
It was intentional. Zach's [User Story](https://ni.visualstudio.com/DevCentral/_workitems/edit/2182524) for this said: "It is only necessary for the Init calls, but if that is hard to scope to just those methods, all functions are fine."

The API key is indeed just for licensing -- I explicitly avoid using "auth" or "authentication" in any of my naming. Authentication would happen at the base gRPC layer, typically via SSL/TLS.

Reply:
> Authentication would happen at the base gRPC layer, typically via SSL/TLS.

FYI, with server-side TLS, it is common for the client to send authentication headers (such as OAuth2 tokens) with each HTTPS request.

Reply:
At the moment, we only foresee needing it during init, so I'm going to avoid adding the extra complexity of an interceptor for now, and just leave it as-is.